calendar_resource/models/calendar_event.py

```diff
@@ -0,0 +1,268 @@
+                    (
+                        datetime(2017, 6, 29, 12, 0, 0),
+                        datetime(2017, 6, 29, 13, 0, 0),
+                    ),
+                ],
+            )
+
+            The above example will output the following string,
+            depending on date and time format, as well as timezone.
+
+            \n06/28/2017 at 17:00:00 To\n
+            06/29/2017 at 08:00:00 (UTC)\n
+
+            \n06/29/2017 at 12:00:00 To\n
+            06/29/2017 at 13:00:00 (UTC)\n
+
+        Returns:
+
+            (str): string formatted for front-end
+            display purposes, taking into account language, timezone, as
+            well as date and time formats.
+
+        """
+        datetimes = []
+        for interval in intervals:
+            if not isinstance(interval[0], str):
+                interval = (
+                    fields.Datetime.to_string(interval[0]),
+                    fields.Datetime.to_string(interval[1]),
+                )
+            args = {
+                'start': interval[0],
+                'stop': interval[1],
+                'zallday': False,
+                'zduration': 24,
+            }
+            datetimes.append(self._get_display_time(**args))
+        return (2 * linesep).join(datetimes)
+
+    @api.multi
+    def _event_in_past(self):
+        self.ensure_one()
+        stop_datetime = fields.Datetime.from_string(self.stop)
+        now_datetime = datetime.now()
+        return stop_datetime < now_datetime
+
+    @api.multi
+    @api.constrains('resource_ids', 'start', 'stop')
+    def _check_resource_ids_double_book(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            resources = record.resource_ids.filtered(
+                lambda s: s.allow_double_book is False
+            )
+
+            if not any(resources):
+                continue
+
+            overlaps = self.env['calendar.event'].search([
+                ('id', '!=', record.id),
+                ('start', '<', record.stop),
+                ('stop', '>', record.start),
+            ])
+            for resource in overlaps.mapped(lambda s: s.resource_ids):
+                raise ValidationError(
+                    _(
+                        'The resource, %s, cannot be double-booked '
+                        'with any overlapping meetings or events.',
+                    )
+                    % resource.name,
+                )
+
+    @api.multi
+    @api.constrains('resource_ids', 'categ_ids')
+    def _check_resource_ids_categ_ids(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            if not record.categ_ids:
+                continue
+
+            for resource in record.resource_ids:
+                categs = record.categ_ids.filtered(
+                    lambda s: s not in resource.allowed_event_types
+                )
+                if categs:
```

Comment:
Replace this for by a search with limit = 1, similar to the above commented.

Reply:
Right, thanks